### PR TITLE
Services_Stack

### PR DESCRIPTION
**Pull request by mickaelgadoud**
### Pre-review TODO

- [x] PR is rebased on top of `origin/master` 
- [ ] PR provide new tests for new behaviors
- [ ] PR doesn't introduce commented out code
- [ ] PR has been tested locally on the main use cases
- [ ] Documentation have been updated if necessary

### CODE REVIEW

File: server/app/controllers/service-controller.js [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+ * Create a service.
+ *
+ * @param req Request
+ * @param res Response
+ * @return {Promise.<void>} Nothing
+ */
+async function createService(req, res) {
+
+    // FIXME We should check the type of each provided field, instead of just the presence
+    if (!checkStructure(req.body, ['name', 'startAt', 'endAt', 'nbMax', 'category'])) {
```
Comment (kistora) [inline server/app/controllers/service-controller.js]: `name` field is not required

File: server/app/controllers/service-controller.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        throw createUserError(
+            'BadRequest',
+            'The body has missing properties, needed: [\'name\', \'startAt\', \'endAt\', \'nbMax\', \'category\']'
+        );
+    }
+
+    let newService = new Service({
+        name: req.body.name,
```
Comment (kistora) [inline server/app/controllers/service-controller.js]: `name: req.body.name || "",`

File: server/app/controllers/service-controller.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ * Update a service.
+ *
+ * @param req Request
+ * @param res Response
+ * @return {Promise.<void>} Nothing
+ */
+async function updateService(req, res) {
+
+    // FIXME We should check the type of each provided field, instead of just the presence
+    if (!checkStructure(req.body, ['name', 'startAt', 'endAt', 'nbMax', 'category'])) {
```
Comment (kistora) [inline server/app/controllers/service-controller.js]: Same here, `name` field is not required

File: server/app/controllers/service-controller.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        throw createUserError(
+            'BadRequest',
+            'The body has missing properties, needed: [\'name\', \'startAt\', \'endAt\', \'nbMax\', \'category\']'
+        );
+    }
+
+    let newService = new Service({
+        name: req.body.name,
```
Comment (kistora) [inline server/app/controllers/service-controller.js]: Same

File: server/app/controllers/service-controller.js [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+
+module.exports = {
+    getAllServices,
+    createService,
+    updateService,
+    getServiceById,
+    deleteService
+};
```
Comment (kistora) [inline server/app/controllers/service-controller.js]: Miss empty line (required for ESLint tests)

File: server/app/routes/services.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+
+router.get('/', guard.check('service:read'), am(serviceController.getAllServices));
+
+router.post('/', am(serviceController.createService));
+
+router.get('/:id(\\d+)', am(serviceController.getServiceById));
+router.put('/:id(\\d+)', am(serviceController.updateService));
+router.delete('/:id(\\d+)', am(serviceController.deleteService));
+
+module.exports = router;
```
Comment (kistora) [inline server/app/routes/services.js]: Miss empty line

File: server/app/routes/services.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+const router = require('express').Router();
+const guard = require('express-jwt-permissions')();
+const am = require('../../utils/async-middleware');
+const serviceController = require('../controllers/service-controller');
+
+router.get('/', guard.check('service:read'), am(serviceController.getAllServices));
+
```
Comment (kistora) [inline server/app/routes/services.js]: Useless line

File: server/app/routes/services.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+router.post('/', am(serviceController.createService));
+
```
Comment (kistora) [inline server/app/routes/services.js]: Useless line again :)

File: server/app/controllers/service-controller.js (hunk 2 above)
Comment (Embraser01) [reply server/app/controllers/service-controller.js]: Not necessary now :smiley:

File: server/app/controllers/service-controller.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+/**
+ * Create a service.
+ *
+ * @param req Request
+ * @param res Response
+ * @return {Promise.<void>} Nothing
+ */
+async function createService(req, res) {
```
Comment (Embraser01) [inline server/app/controllers/service-controller.js]: Actually, we decided to create an array of services, see swagger for more details.

Comment (Embraser01) [reply server/app/controllers/service-controller.js]: We will do this in another PR

**Pull request merged**